Makes access_network more available

mikomyazaki opened the pull request:
:cl: mikomyazaki
tweak: Senior Engineer now gets access_network, allowing them to use the NTNet Diagnostics and Monitoring program and various command line tools.
tweak: Access hacking antag program can now hack access_network, if the antag has master level IT skill.
tweak: The NTNet Quantum Relay in telecomms will now keep physical logs of network activity, wiping the NTNet Diagnostics Program logs will not affect these, and they cannot be accessed remotely. The portable drive can be removed via a UI eject button and the data imported from it into a modular computer.
maptweak: Telecomms storage room now has a couple of spare portable hard disks.
/:cl:

access_network is currently only available to the CO, XO, CE and AI/borgs, despite having many uses that are solely for antagonists (proxies etc.)

This change makes it so that antagonists with high IT skill and an emag can access the hacking minigame, which can lead to interesting times for Forensics + Engineering trying to chase them down.

I've also given access_network to the Senior Engineer to increase the chances there is someone able to oppose an antagonist who uses the hacking features. The Senior Engineer already has access to telecomms.

Physical logs will require the antagonist to do more to cover their tracks, and they'll probably leave some fingerprints and fibers around unless they're very careful.

(I recently wrote a wiki guide on how to use the network hacking tools and defend against them, so while this is a kind of rarely used feature, I hope people will be able to work it out easily as it becomes a more commonly used tool by antagonists.)

mikomyazaki commented:
Travis fails are due to docking controller weirdness, that is not touched by this PR.

mikomyazaki commented:
Following feedback, will be doing the following changes:
- [x] ID card access readded to the blacklist.
- [x] NTNet Diagnostics program will make a physical log file on the telecomms computer, so the antag will need to deal with that if they want to erase all traces.
- [x] Will add a skill requirement to hack access_network (Master IT)

mikomyazaki commented:
Made all the changes based on feedback, ready for review now.

comma commented on code/modules/modular_computers/NTNet/NTNet.dm:
```diff
@@ -64,6 +66,15 @@ var/global/datum/ntnet/ntnet_global = new()
 			else
 				break
 
+	for(var/obj/item/modular_computer/console/preset/ntnet_mon/C in GLOB.ntnet_physical_log_computers)
```
That's pretty hecking gross, means can't build them for example, and perversing what presets are for tbh

mikomyazaki replied:
Yeah. I'm going to stick this function into the NTNet Relay with some sort of removable hard-disk drive. Should be nicer.

comma reviewed (dismissed):
Access stuff is fine, but logging should be redone to not use presets

mikomyazaki commented:
Changed things so it writes logs to an internal HDD in the NTNet Quantum Relay. UI now looks like this:
![new_ui1](https://user-images.githubusercontent.com/47489928/61706279-8fe1de00-ad3f-11e9-8872-14b032c1902a.PNG)
![new_ui2](https://user-images.githubusercontent.com/47489928/61706280-92443800-ad3f-11e9-83f3-87adaf5e7f4f.PNG)

Mapped in a couple of spare portable hard drives to the telecomms storage area.

comma commented on code/modules/modular_computers/NTNet/NTNet_relay.dm:
```diff
@@ -102,6 +109,7 @@
 		ntnet_global.relays.Add(src)
 		NTNet = ntnet_global
 		ntnet_global.add_log("New quantum relay activated. Current amount of linked relays: [NTNet.relays.len]")
+	portable_drive = new/obj/item/weapon/stock_parts/computer/hard_drive/portable(src)
```
SHould probably do this earlier in the proc, since there's add_log call just above :V

mikomyazaki replied:
Moved it to the start of New().

comma commented on code/modules/modular_computers/NTNet/NTNet_relay.dm:
```diff
@@ -112,4 +120,14 @@
+	else if(user.unEquip(P,src))
+		portable_drive = P
+		to_chat(user, "You install \the [portable_drive] into \the [src]")
+		portable_drive.holder2 = src
```
holder2 isn't cleared when you eject disk, unsure if it even needs to be set but need to check disk code for that

mikomyazaki replied:
I don't think it does do anything, but holder2 *is* set for every other computer component when it gets put into something. So maybe it should have it just incase that matters for future functionality?

I've made the var get cleared when ejecting.

PsiOmegaDelta commented on code/modules/modular_computers/NTNet/NTNet_relay.dm:
```diff
@@ -15,6 +15,7 @@
 	var/enabled = 1				// Set to 0 if the relay was turned off
 	var/dos_failure = 0			// Set to 1 if the relay failed due to (D)DoS attack
 	var/list/dos_sources = list()	// Backwards reference for qdel() stuff
+	var/obj/item/weapon/stock_parts/computer/hard_drive/portable/portable_drive              // drive for logs storage
```
Need to handle this reference in `/Destroy()`

afterthought2 replied:
It shouldn't actually be here. Install it as a component (note how it has the correct type path for it). You can then fetch it with `get_component_of_type`, and everything else is handled internally.

mikomyazaki replied:
Hopefully resolved now?

mikomyazaki replied:
I'm not sure on the proper place to put `portable_drive = get_component_of_type(drivetype)`

It works where it is now, in Process, but I doubt that is the proper place to have it. Also it doesn't log its own creation, but it doesn't work properly at the start of New().

mikomyazaki replied:
This has been changed to use the getter rather than a local var, so marking this as resolved

PsiOmegaDelta commented on code/modules/modular_computers/NTNet/NTNet_relay.dm:
```diff
@@ -112,4 +121,14 @@
+	else if(user.unEquip(P,src))
+		portable_drive = P
+		to_chat(user, "You install \the [portable_drive] into \the [src]")
+		portable_drive.holder2 = src
```
This `holder2` thing seems like something that should be set/reset via procs on `portable_drive`

mikomyazaki replied:
I think it isn't needed for the NTNet Quantum Relay. It gets handled already when we put it in a modular computer to read the data. So I'll leave it out unless @afterthought2 thinks I need it.

afterthought2 replied:
It's incorrect to set holder2 at all, most likely; it expects a modular computer, not an arbitrary obj.

mikomyazaki replied:
Should be resolved now. Removed holder2 stuff.

PsiOmegaDelta commented on code/modules/modular_computers/file_system/programs/antagonist/access_decrypter.dm:
```diff
@@ -79,12 +83,14 @@
 			return 1
 		if(access in restricted_access_codes)
 			return 1
+		if(access in skill_restricted_access_codes_master && operator_skill < SKILL_PROF)
```
The `in` operator has a rather weak operator precedence and should have extra brackets:
`if((access in skill_restricted_access_codes_master) && operator_skill < SKILL_PROF)`

mikomyazaki replied:
Done.

PsiOmegaDelta commented on code/modules/modular_computers/NTNet/NTNet_relay.dm:
```diff
@@ -64,6 +65,7 @@
 	data["dos_capacity"] = dos_capacity
 	data["dos_overload"] = dos_overload
 	data["dos_crashed"] = dos_failure
+	data["portable_drive"] = portable_drive
```
Mainly noting that a more appropriate setup would be:
`data["has_portable_drive"] = !!portable_drive`

mikomyazaki replied:
Done.

mikomyazaki commented:
I can't test these last changes for three days due to being away from home. Could you mark it as DNM until I can do that?

mikomyazaki commented:
Tested and working, some possible issues I've commented above. Pls rereview, thanks.

MistakeNot4892 commented:
For reference you can get the bot to mark it DNM by editing the title to have [DNM] at the start.

MistakeNot4892 commented on code/modules/modular_computers/NTNet/NTNet.dm:
```diff
@@ -64,6 +64,16 @@ var/global/datum/ntnet/ntnet_global = new()
 				break
 
+	for(var/obj/machinery/ntnet_relay/R in ntnet_global.relays)
+		if(R?.portable_drive)
```
Null coalescence shouldn't be necessary in a typed loop as `R` will be null and typechecked by BYOND before you get to this point.

MistakeNot4892 commented on code/modules/modular_computers/NTNet/NTNet_relay.dm:
```diff
@@ -43,6 +44,8 @@
 		update_use_power(POWER_USE_ACTIVE)
 	else
 		update_use_power(POWER_USE_IDLE)
+	if(!portable_drive)
```
Should this be in `Initialize()` instead? I'm not overly familiar with how these machines work internally.

afterthought2 replied:
Shouldn't need a local var for it. Just call the getter if needed.

mikomyazaki replied:
Changed things to use get_component_of_type instead of a local var.

MistakeNot4892 commented on code/modules/modular_computers/NTNet/NTNet_relay.dm:
```diff
@@ -94,6 +98,10 @@
 		ntnet_global.banned_nids.Cut()
 		ntnet_global.add_log("Manual override: Network blacklist cleared.")
 		return 1
+	else if(href_list["eject_drive"])
```
Minor quibble, you could collapse the following `if` into this line.

MistakeNot4892 reviewed (commented):
I'm not familiar enough with the system to give an approval but the code itself largely looks fine outside of my comments.

mikomyazaki committed:
SE gets access_network
Antag access hacking program can get access_network with master IT
NTNet Quantum Relay keeps logs, has a portable hard drive that can be removed for reading
Maps in a couple of extra portable hard drives to the telecomms storage
